### ISSUE TITLE
Stale `Cabal-simple` file

### ISSUE THREAD

**Issue by marcosh** [commercialhaskell/stack#5829]
Trying to compile a project with `Stack` inside a `nix-shell`, I got the following error

```
/home/marcosh/.stack/setup-exe-cache/x86_64-linux-nix/Cabal-simple_mPHDZzAJ_3.4.1.0_ghc-9.0.2: startProcess: exec: invalid argument (Bad file descriptor)
```

The problem seems to happen because the file refers to the executables from my system

```
> ldd Cabal-simple_mPHDZzAJ_3.4.1.0_ghc-9.0.2 
        linux-vdso.so.1 => linux-vdso.so.1 (0x00007ffdb0521000)
        libgmp.so.10 => /usr/lib/libgmp.so.10 (0x00007fe0b166a000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007fe0b1486000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007fe0b13a0000)
        librt.so.1 => /usr/lib/librt.so.1 (0x00007fe0b1399000)
        libdl.so.2 => /usr/lib/libdl.so.2 (0x00007fe0b1394000)
        libffi.so.8 => /usr/lib/libffi.so.8 (0x00007fe0b1388000)
        /nix/store/bvy2z17rzlvkx2sj7fy99ajm853yv898-glibc-2.34-210/lib/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007fe0b173c000)
```

Removing that file, compilation works, the file is regenerated with executables actually coming from nix

```
> ldd Cabal-simple_mPHDZzAJ_3.4.1.0_ghc-9.0.2 
        linux-vdso.so.1 => linux-vdso.so.1 (0x00007fff15969000)
        libgmp.so.10 => /nix/store/4l48ybgm682wr7s2m1bzr85ca51i3kgb-gmp-with-cxx-6.2.1/lib/libgmp.so.10 (0x00007f968da9a000)
        libc.so.6 => /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/libc.so.6 (0x00007f968d89c000)
        libm.so.6 => /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/libm.so.6 (0x00007f968d7c3000)
        librt.so.1 => /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/librt.so.1 (0x00007f968d7bc000)
        libdl.so.2 => /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/libdl.so.2 (0x00007f968d7b7000)
        libffi.so.8 => /nix/store/7m5ik9rhh55m021kzmg4hb4zjfj3lyyc-libffi-3.4.2/lib/libffi.so.8 (0x00007f968d7aa000)
        /nix/store/lyl6nysc3i3aqhj6shizjgj0ibnf1pvg-glibc-2.34-210/lib/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f968db3d000)

**Comment by mpilgrem** [comment]
Thanks. I am a Windows user, so I cannot help but I've labelled this as Nix-related to bring it to the attention of others interested in Stack's Nix integration. May I ask, do you know what might have caused the original:
~~~
x86_64-linux-nix/Cabal-simple_mPHDZzAJ_3.4.1.0_ghc-9.0.2.exe
~~~
to have been created?

I am wondering if the (intermediate) fix for this is to add something to the online help at: https://docs.haskellstack.org/en/stable/nix_integration/

**Comment by marcosh** [comment]
probably it was there from a previous attempt at using nix for building the project. Not really sure though

**Comment by tomjaguarpaw** [comment]
I don't think this is due to Nix. I also just got this error on the `typed-process` Github Actions CI setup when adding LTS-18 to LTS-21 support. The failures where on MacOS only.

```
      /Users/runner/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_6HauvNHV_2.4.0.1_ghc-8.6.5: startProcess: exec: invalid argument (Bad file descriptor)
```

https://github.com/tomjaguarpaw/haskell-opaleye/actions/runs/6423951056/job/17495833385

I cleared the Github Actions cache and the problem seems to have gone but I wonder if there's some bad caching behaviour on the part of Stack that keeps these files around when it shouldn't.

**Comment by tomjaguarpaw** [comment]
In fact this is really bad, because it seems that when I run with a clean cache there is no problem, but subsequently I get the above error

https://github.com/fpco/typed-process/actions/runs/6443636657/job/17496198334

**Comment by tomjaguarpaw** [comment]
This is quite hard to debug because it has to go through a Github Actions CI loop each time I want to make a change. It turns out the file `Cabal-simple_6HauvNHV_2.4.0.1_ghc-8.6.5` doesn't actually exist. Why would `stack` think it exists? And if it doesn't exist, why wouldn't it try to rebuild it? Is there any light you can shed here @mpilgrem, because I'm rather lost.

**Comment by tomjaguarpaw** [comment]
I managed to fix the issue by using a newer Github Actions cache action. I still don't understand why the file in question was missing from the cache or why Stack didn't rebuild it when it noticed it was missing, but at least everything is working fine for me now.

**Comment by mpilgrem** [comment]
@tomjaguarpaw, I'll look into this. As to what that executable is, see the (recently added) online documentation at: https://docs.haskellstack.org/en/stable/stack_root/#setup-exe-cache-directory.

**Comment by tomjaguarpaw** [comment]
Thanks! The behaviour I'm seeing is quite strange. Is `setup-exe-cache/x86_64-osx/Cabal-simple_6HauvNHV_2.4.0.1_ghc-8.6.5` by any chance expected to be a symlink? That seems the most plausible explanation for what's going on.

**Comment by mpilgrem** [comment]
@tomjaguarpaw, it should not be a symlink. It should be an executable built from corresponding source code in `setup-exe-src`. The source code has changed in recent version of Stack. I need to check that will not have confused things elsewhere.
